Add support for transient textures on Vulkan and Metal

opstic opened the pull request:
**Connections**
Fixes #8239

**Description**
Exposes `MTLStorageMode.memoryless` for Metal and `vk::MemoryPropertyFlags::LAZILY_ALLOCATED` for Vulkan by adding `TextureUsages::TRANSIENT`.

**Testing**
Added corresponding `TextureUsage` flag to relevant examples.
Also added some render tests and validation testing.

**Squash or Rebase?**
Squash

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

opstic committed:
Preliminary implementation

opstic committed:
Merge branch 'gfx-rs:trunk' into transient

opstic committed:
oops

opstic committed:
Better implementation

opstic committed:
Add entry in `CHANGELOG.md`

cwfitzgerald reviewed (changes requested):
Thanks for taking this on! It just so happened we just talked about wanting this in the WebGPU face to face, so super happy to have this! There's a few things that need to be added, but I don't think it will end up being too much work:

1. At the face to face, we've decided that transient textures wouldn't be a feature. It would be available everywhere, but a no-op on unsupported platforms. If we wanted to expose to the user if this is supported, we can do this in the AdapterInfo as a hint.
2. We need two pieces of validation:
  a. If you add the transient attachment usage, the usage must be exactly RENDER_ATTACHMENT | TRANSIENT
  b. When writing a renderpass to these attachments, the StoreOp must be discard.
3. With point 1 implemented, we should stick these flags on as many of the examples as relevant as we really want to encourage its use.
4. On VK, we need to add `VK_MEMORY_PROPERTY_LAZILY_ALLOCATED_BIT` to the memory so that it may be lazily allocated. I think you may be getting this already implicitly without explicitly asking for it, if you're seeing memory improvements.
5. We should add some tests
  a. For all the validation I mentioned, these can be simple tests in wgpu-validation.
  b. We should have at least one test in the wgpu-gpu tests which ensures we don't get validation errors from vk/metal.

Let me know if you have any questions or if any of this is unclear!

cwfitzgerald commented on wgpu-hal/src/vulkan/device.rs:
```diff
@@ -1264,13 +1264,19 @@ impl crate::Device for super::Device {
             })?;
 
+        let mut alloc_usage = gpu_alloc::UsageFlags::FAST_DEVICE_ACCESS;
+        alloc_usage.set(
+            gpu_alloc::UsageFlags::TRANSIENT,
+            desc.usage.contains(wgt::TextureUses::TRANSIENT),
+        );
```
This is, confusingly, an unrelated use of the word transient - this is for very short lived buffers and stuff. It uses a different allocation scheme, unrelated to transient textures.

opstic replied:
Redone it the proper way by specifying a `MemoryType` of `LAZILY_ALLOCATED` if found. Thanks!

opstic committed:
Remove `Feature::TRANSIENT_ATTACHMENTS` and add validation

opstic committed:
Remove `#[cfg(windows)]` from `find_memory_type_index`

opstic committed:
Tries to find a memory type that contains `MemoryPropertyFlags::LAZILY_ALLOCATED` and applies it

opstic committed:
Merge branch 'trunk' into transient

opstic commented:
Thanks for your input @cwfitzgerald!

For:

2. I added some validation and error types specific to the situations you mentioned, please check it again!

3. The most relevant looks to only be the `msaa_line` example, I looked through all of the other examples and it appears they don't use multisampling at all. I'll be adding it now.

4. Something I found was that none of the desktop GPUs I have supports memory types with `LAZILY_ALLOCATED`, the memory decrease must just have been run-to-run difference😭, I'll be proceeding by testing more carefully and on mobile from now on.

opstic committed:
`cargo fmt`

opstic committed:
Add check in Metal for `MTLStorageMode::Memoryless` support

cwfitzgerald commented:
Re 3: `shadow` can use it for the forward depth buffer, and I suspect there might be an extra example or two which only ever uses depth in a memoryless compatible way.

cwfitzgerald commented:
> Something I found was that none of the desktop GPUs I have supports memory types with LAZILY_ALLOCATED, the memory decrease must just have been run-to-run difference😭, I'll be proceeding by testing more carefully and on mobile from now on.

Yeah this is unfortunately a mobile only feature really, but hopefully the validation layers can save us on desktop. Apple is the main one that really benefits from it on "desktop".

cwfitzgerald commented on wgpu-core/src/command/render.rs:
```diff
@@ -625,6 +625,8 @@ pub enum ColorAttachmentError {
         mip_level: u32,
         depth_or_array_layer: u32,
     },
+    #[error("Color attachment's usage cannot contain {0:?} if StoreOp is {1:?}")]
```
It would be a bit more helpful to the user if this said something like this. Maybe make the middle "StoreOp::discard" an argument too? Not sure.

```suggestion
    #[error("Color attachment's usage contains {0:?}. This can only be used with StoreOp::Discard, but the provided store op was {1:?}")]
```

opstic commented:
> Yeah this is unfortunately a mobile only feature really, but hopefully the validation layers can save us on desktop. Apple is the main one that really benefits from it on "desktop".

I did make it so that a if a memory type with the `LAZILY_ALLOCATED` bit was not found, it would just proceed normally with the `TextureUsages::TRANSIENT` being disregarded, just making sure that this behavior should be the intended one.

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -5610,6 +5610,8 @@ bitflags::bitflags! {
         //
         /// Allows a texture to be used with image atomics. Requires [`Features::TEXTURE_ATOMIC`].
         const STORAGE_ATOMIC = 1 << 16;
+        /// Allows a texture to be transient. No-op on platforms other than Vulkan and Metal.
```
This should include the fact that it is incompatible with every other usage except render attachment. Maybe also add a note to render attachment explaining when you should also enable transient.

cwfitzgerald replied:
Oh, it should also include more of a description of what this is and why you'd use it, noting that there's no harm in adding it if it's compatible with your workflow.

opstic replied:
Done! Please check if there's anything else I need to mention in the documentation.

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -5664,6 +5666,8 @@ bitflags::bitflags! {
         /// Flag used by the wgpu-core texture tracker to say that the tracker does not know the state of the sub-resource.
         /// This is different from UNINITIALIZED as that says the tracker does know, but the texture has not been initialized.
         const UNKNOWN = 1 << 13;
+        /// Transient image.
```
Lets put this after STORAGE_ATOMIC at 1 << 12, just for grouping purposes, kicking the rest down.

```suggestion
        /// Transient texture that may not have any backing memory. Not a resource state stored in the trackers, only used for passing down usages to create_texture.
```

cwfitzgerald reviewed (changes requested):
Some comments and needs tests, but moving in the right direction!

opstic commented:
> Re 3: shadow can use it for the forward depth buffer, and I suspect there might be an extra example or two which only ever uses depth in a memoryless compatible way.

Alright, I looked through it more carefully again, we should be able to apply `TRANSIENT` on `msaa_line`'s resolve texture and `shadow`, `skybox`'s depth texture. Adding them now.

opstic committed:
Apply `TextureUsages::TRANSIENT` to relevant examples

opstic committed:
Improve documentation by @cwfitzgerald

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

opstic committed:
Improve error message by @cwfitzgerald

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

opstic committed:
Move `TextureUses::TRANSIENT` to `1 << 12`

opstic committed:
Improve `TextureUsages` docs

opstic committed:
Better wording

opstic committed:
Add `StoreOp` requirement

opstic committed:
Allow `TextureUsages::TRANSIENT` in `guaranteed_format_features`

opstic committed:
Improve error message

opstic committed:
Add `TextureUses::TRANSIENT` to GL backend's `render_usage`

opstic committed:
Improve docs

opstic committed:
Add validation tests

opstic commented:
Hi @cwfitzgerald!

I have done most of your suggestions now, there are just two things that I'm not sure about:

1. Not really sure how to add the hint for `TRANSIENT` support, just add a bool field in `AdapterInfo`?
2. First time contributing to wgpu, thus not too sure on what the tests in wgpu-gpu should look like. Would adding a test that does one render pass with a `TRANSIENT` texture work?

Thanks again!

cwfitzgerald commented:
> Not really sure how to add the hint for TRANSIENT support, just add a bool field in AdapterInfo?

Yeah that's fine, we don't really have a precedent for this, so we can just kinda make it up.

> First time contributing to wgpu, thus not too sure on what the tests in wgpu-gpu should look like. Would adding a test that does one render pass with a TRANSIENT texture work?

Yup! Just to make sure there's nothing terribly wrong going on.

opstic committed:
Fix `TextureUsages::TRANSIENT` docs formatting

opstic committed:
Add `supports_transient` flag in `AdapterInfo`

opstic committed:
Merge branch 'trunk' into transient

opstic committed:
oops

opstic committed:
📎

opstic committed:
Formatting

opstic committed:
Fix WebGPU compile

opstic committed:
Add simple transient test case

Renders a single triangle with transient texture as target to make sure nothing is wrong

opstic committed:
Actually use the transient flag in test

opstic committed:
Merge branch 'trunk' into transient

opstic committed:
Well that was a nice demonstration of validation

cwfitzgerald commented:
Confirmed that this is working as expected on mac - will do another review.

opstic committed:
Update `CHANGELOG.md`

cwfitzgerald commented:
(I'm doing this to see how it does, I'll do a full human review tomorrow, no action is needed based on its comments).

cwfitzgerald commented on tests/tests/wgpu-validation/api/texture.rs:
```diff
@@ -508,3 +508,96 @@ fn copy_buffer_to_texture_forbidden_format_aspect() {
+            || {
+                device.create_texture(&invalid_texture_descriptor)
+            },
+            Some(&format!("Texture usage TextureUsages(TRANSIENT) is not compatible with texture usage {usage:?}")),
+        );
```
No lol

opstic commented on wgpu-types/src/lib.rs:
```diff
@@ -5601,6 +5605,8 @@ bitflags::bitflags! {
         const STORAGE_BINDING = 1 << 3;
         /// Allows a texture to be an output attachment of a render pass.
+        ///
+        /// Consider adding ['TextureUsages::TRANSIENT'] if the contents are not reused.
```
This thing actually caught something I've missed, surprisingly.

opstic committed:
Use proper formatting for documentation

opstic committed:
Merge branch 'trunk' into transient

opstic committed:
Merge branch 'trunk' into transient

cwfitzgerald commented on tests/tests/wgpu-validation/api/texture.rs:
```diff
@@ -508,3 +508,96 @@ fn copy_buffer_to_texture_forbidden_format_aspect() {
+        depth_or_array_layers: 1,
+    };
+
+    let invalid_usages = [
+        wgpu::TextureUsages::COPY_SRC,
+        wgpu::TextureUsages::COPY_DST,
+        wgpu::TextureUsages::TEXTURE_BINDING,
+        wgpu::TextureUsages::STORAGE_BINDING,
+        wgpu::TextureUsages::STORAGE_ATOMIC,
+    ];
```
I would define this as 

```suggestion
    let invalid_usages = wgpu::TextureUsages::all() - wgpu::TextureUsages::RENDER_ATTACHMENT - wgpu::TextureUsages::TRANSIENT;
```

so that it automatically expands if we add more usages.

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -3253,6 +3255,8 @@ impl TextureFormat {
             allowed_usages.contains(TextureUsages::STORAGE_ATOMIC),
         );
 
+        allowed_usages |= TextureUsages::TRANSIENT;
```
Hmm I don't think this is actually true - I think it needs to be only the textures which have RENDER_ATTACHMENT.

You can include this in the `attachment` variable.

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -5610,6 +5616,14 @@ bitflags::bitflags! {
         const STORAGE_ATOMIC = 1 << 16;
+        /// Specifies the contents of this texture will not be used in another pass to potentially reduce memory usage and bandwidth.
+        ///
+        /// No-op on platforms other than Vulkan and Metal.
```
```suggestion
        /// No-op on platforms on platforms that do not benefit from transient textures.
        /// Generally mobile and Apple chips care about this.
```

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -1414,6 +1414,8 @@ pub struct AdapterInfo {
     pub driver_info: String,
     /// Backend used for device
     pub backend: Backend,
+    /// Supports [`TextureUsages::TRANSIENT`]
```
```suggestion
    /// If true, adding [`TextureUsages::TRANSIENT`] will decrease memory usage.
```

cwfitzgerald commented on wgpu-hal/src/noop/mod.rs:
```diff
@@ -140,6 +140,7 @@ pub fn adapter_info() -> wgt::AdapterInfo {
         driver: String::from("wgpu"),
         driver_info: String::new(),
         backend: wgt::Backend::Noop,
+        supports_transient: true,
```
```suggestion
        supports_transient: false,
```

See the change in the doc comment on this field for the doc comment

cwfitzgerald replied:
This needs to be updated here too

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -1414,6 +1414,8 @@ pub struct AdapterInfo {
     /// Backend used for device
     pub backend: Backend,
+    /// Supports [`TextureUsages::TRANSIENT`]
+    pub supports_transient: bool,
```
```suggestion
    pub native_transient_texture: bool,
```

Really not set on this name, but what I'm trying to avoid is the implication that TRANSIENT is not allowed

cwfitzgerald replied:
This still needs to be updated

opstic replied:
Oops, been pretty busy recently. I should be able to finish the rest of the changes this weekend. Sorry for the wait.

cwfitzgerald replied:
No worries or rush! Just re-request a review from me whenever you're ready!

opstic replied:
Changed it to `transient_saves_memory`, please review if that name looks good or if I should change it to anything else!

cwfitzgerald reviewed (changes requested):
Looking good, I thought of one more potential problem:

We need to validate out creating a texture with _only_ TRANSIENT usage. I'm not sure what metal would do with this, but it can't be good 😆 

One final thing, you should do a pass over the backends and make sure transient is advertised https://github.com/gfx-rs/wgpu/blob/trunk/wgpu-hal/src/lib.rs#L670-L673 wherever RENDER_ATTACHMENT is advertised.

Thanks for being so helpful with this process, we should be almost finished!!

opstic committed:
Improve validation usage filtering by @cwfitzgerald

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

opstic committed:
Improve documentation by @cwfitzgerald

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

opstic committed:
Improve documentation by @cwfitzgerald

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

opstic committed:
Limit `supports_memoryless_storage` to Apple Silicon

opstic committed:
Merge remote-tracking branch 'fork/transient' into transient

opstic committed:
`cargo fmt`

opstic committed:
Move `TextureUsages::TRANSIENT` in `guaranteed_format_features`

opstic committed:
Merge branch 'trunk' into transient

opstic committed:
Remove unnecessary `mut`

opstic committed:
`MTLGPUFamily::Metal4` doesn't exist yet in the version on `metal` we're using

JMS55 commented:
What's the use case for this? 

One thing I do in Bevy is use a dummy render target for a render pass, since I use storage atomics for my output. I have wanted render passes with no render attachments for this reason, so that we don't need to waste memory on a dummy render target. Would this be a suitable replacement?

cwfitzgerald commented:
> What's the use case for this?

Tilers where you need an attachment but never actually read from it from memory. You can avoid allocating that memory.

> One thing I do in Bevy is use a dummy render target for a render pass, since I use storage atomics for my output. I have wanted render passes with no render attachments for this reason, so that we don't need to waste memory on a dummy render target. Would this be a suitable replacement?

This is better than nothing, but not really. It would only actually reduce memory on tilers.

opstic committed:
Improve transient support hint naming

opstic committed:
Correct `AdapterInfo` for `noop` backend

opstic committed:
Merge branch 'trunk' into transient

opstic committed:
Fix compile

opstic committed:
`cargo fmt`

opstic committed:
Merge branch 'trunk' into transient

opstic committed:
Allow `TextureUsages::TRANSIENT` along with `TextureUsages::RENDER_ATTACHMENT`

opstic committed:
Disallow `TextureUsages::TRANSIENT` if `TextureUsages::RENDER_ATTACHMENT` isn't present.

opstic committed:
Update docs

opstic commented:
> One final thing, you should do a pass over the backends and make sure transient is advertised https://github.com/gfx-rs/wgpu/blob/trunk/wgpu-hal/src/lib.rs#L670-L673 wherever RENDER_ATTACHMENT is advertised.

I checked and it appears the only usage of `texture_format_capabilities` seems to be in [`get_texture_format_features`](https://github.com/gfx-rs/wgpu/blob/8ff1eb995df2662fa6c066c2b07c3cfcd95ea2bb/wgpu-core/src/instance.rs#L677C12-L677C39), which is then used in other places, so I think it should be sufficient to advertise it there.

opstic committed:
Update validation test

opstic committed:
`cargo fmt`

cwfitzgerald commented:
Sure, that works

opstic committed:
Merge branch 'trunk' into transient

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -1415,6 +1415,8 @@ pub struct AdapterInfo {
     pub driver_info: String,
     /// Backend used for device
     pub backend: Backend,
+    /// If true, adding [`TextureUsages::TRANSIENT`] will decrease memory usage.
```
```suggestion
    /// If true, adding [`TextureUsages::TRANSIENT`] to a texture will decrease memory usage.
```

cwfitzgerald reviewed (approved):
Great stuff!

cwfitzgerald committed:
Update wgpu-types/src/lib.rs

inner-daemons commented:
@opstic, FYI your feature detection code for metal

```rust
if family_check {
    device.supports_family(MTLGPUFamily::Apple2)
} else {
    version.at_least((11, 0), (10, 0), os_is_mac)
}
```
is incorrect. The docs for [`MTLStorageMode.memoryless`](https://developer.apple.com/documentation/metal/mtlstoragemode/memoryless) says 
> The memoryless storage mode uses tile memory, and is only available on Apple family GPUs.

I'm not sure what the else statement above is for.

Additionally, the feature tables say Metal4 is required. I'm leaving this comment so you're aware and to make sure you don't know anything I'm not realizing.

opstic commented:
@inner-daemons It is the best I can do for the `metal` crate we're depending on. Which as of the latest released version, doesn't have the `Metal4` variant available [(docs.rs)](https://docs.rs/metal/latest/metal/enum.MTLGPUFamily.html).

As far as the latest families go, `Metal4` was only recently added for iOS, macOS, etc 26+, and no [documentation](https://developer.apple.com/documentation/metal/mtlgpufamily/metal4) is even available on the specific platforms included, compared to [`Metal3`](https://developer.apple.com/documentation/metal/mtlgpufamily/metal3).

> I'm not sure what the else statement above is for.

That else statement is for older iOS devices, since [`MTLGPUFamily`](https://developer.apple.com/documentation/metal/mtlgpufamily) is only available for iOS 13+, and `MTLStorageMode.memoryless` has been available since iOS 10.

In hindsight, something I am slightly confused by is the requirement for the [`Apple2`](https://developer.apple.com/documentation/metal/mtlgpufamily/apple2), since it corresponds to the Apple A8, while memoryless was mentioned with the Apple A7 during [WWDC 2016](https://nonstrict.eu/wwdcindex/wwdc2016/604/). So might have to do a PR later to fix that...

inner-daemons commented:
> @inner-daemons It is the best I can do for the `metal` crate we're depending on. Which as of the latest released version, doesn't have the `Metal4` variant available [(docs.rs)](https://docs.rs/metal/latest/metal/enum.MTLGPUFamily.html).
> 
> As far as the latest families go, `Metal4` was only recently added for iOS, macOS, etc 26+, and no [documentation](https://developer.apple.com/documentation/metal/mtlgpufamily/metal4) is even available on the specific platforms included, compared to [`Metal3`](https://developer.apple.com/documentation/metal/mtlgpufamily/metal3).
> 
> > I'm not sure what the else statement above is for.
> 
> That else statement is for older iOS devices, since [`MTLGPUFamily`](https://developer.apple.com/documentation/metal/mtlgpufamily) is only available for iOS 13+, and `MTLStorageMode.memoryless` has been available since iOS 10.
> 
> In hindsight, something I am slightly confused by is the requirement for the [`Apple2`](https://developer.apple.com/documentation/metal/mtlgpufamily/apple2), since it corresponds to the Apple A8, while memoryless was mentioned with the Apple A7 during [WWDC 2016](https://nonstrict.eu/wwdcindex/wwdc2016/604/). So might have to do a PR later to fix that...

Thanks for responding quickly. I think I have something misunderstood
* You say memoryless is supported from iOS 10 to iOS 12
* But then on iOS 13 devices with e.g. metal 3, it all of a sudden isn't supported??
* Now with metal 4 it will be reintroduced?

I definitely think you understand this better than me, so I want to know what you think the correct check would be. Because I'm trying to clean up the very flawed and very confusing feature detection logic.

inner-daemons commented:
I don't understand, metal4 isn't even out yet on my operating systems, and yet these docs are claiming metal4 support would be required. So if I went with that check we might as well just disable transient textures for a year. Fuck I don't understand this whatsoever.

opstic commented:
@inner-daemons I think you're misunderstanding the feature tables.

The `Metal` and `Apple` families listed aren't conjunctive, they're disjunctive.

Otherwise for example the 64bit integer support would be very confusing as it would require `Metal3`(A14, M1, etc) *and* `Apple3`(A9). 😉

inner-daemons commented:
@opstic That makes some sense, but for mesh shaders, it seems that they are somehow conjunctive, which is the root of my confusion. Virtual/simulated devices report supporting the right silicon version but only support metal2 and don't support mesh shaders, despite what docs say. Probably that is just poorly written docs.

I need to go to bed, brain broke. Best of luck.

opstic commented:
@inner-daemons ~~As for the mesh shading, the feature tables does indicate support at `Metal3` and/or `Apple7`, so with your problem I think you can do a check just for `Metal3` without needing to check for `Apple7` since `Apple7`(A14) does guarantee `Metal3` support anyways.~~

Edit: Nope scratch that, as `Metal3` is only available on iOS 16+ while `Apple7` is available since iOS 13+, which means that if you check for `Metal3` you'd be removing features for devices with A14 running iOS versions below 16.

inner-daemons commented:
> @inner-daemons As for the mesh shading, the feature tables does indicate support at `Metal3` and/or `Apple7`, so with your problem I think you can do a check just for `Metal3` without needing to check for `Apple7` since `Apple7`(A14) does guarantee `Metal3` support anyways.

In this case, I believe the virtual devices are reporting Apple7 support without Metal3, but still not supporting mesh shaders. I wouldn't like to only check Metal3, because that would eliminate some Mac2 devices. I will figure this out later, thanks for looking at this.
